File: backend/app/schemas/finding_hierarchy.py
"""Read models for the non-destructive, entity-centred Findings workspace."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.models import FindingExclusion, FindingPhase, FindingStatus, Severity
from app.schemas.finding import MAX_FINDING_SUMMARY_CHARS, FindingRead

FindingWorkspaceBucket = Literal[
    "needs_review",
    "actionable",
    "inventory",
    "resolved_excluded",
]
FindingHierarchyKind = Literal[
    "ip",
    "domain",
    "service",
    "subdomain",
    "web_surface",
    "finding",
    "other",
]


class FindingHierarchyFindingRef(BaseModel):
    id: UUID
    title: str
    tool: str | None = None
    target: str | None = None
    severity: Severity
    phase: FindingPhase
    status: FindingStatus
    exclusion: FindingExclusion | None = None
    observed_at: datetime | None = None
    created_at: datetime
    bucket: FindingWorkspaceBucket


class FindingHierarchyRollup(BaseModel):
    max_severity: Severity = Severity.info
    needs_review: int = 0
    actionable: int = 0
    inventory: int = 0
    resolved_excluded: int = 0
    distinct_findings: int = 0
    latest_at: datetime | None = None


class FindingHierarchyItem(BaseModel):
    id: str
    kind: FindingHierarchyKind
    canonical_key: str
    label: str
    value: str | None = None
    ip: str | None = None
    hostname: str | None = None
    protocol: str | None = None
    port: int | None = None
    service: str | None = None
    url: str | None = None
    finding_refs: list[FindingHierarchyFindingRef] = Field(default_factory=list)
    children: list[FindingHierarchyItem] = Field(default_factory=list)
    rollup: FindingHierarchyRollup = Field(default_factory=FindingHierarchyRollup)
    create_finding_allowed: bool = True
    suggested_title: str | None = None
    suggested_target: str | None = None


class FindingHierarchyCounts(BaseModel):
    focus: int = 0
    needs_review: int = 0
    actionable: int = 0
    inventory: int = 0
    resolved_excluded: int = 0
    distinct_findings: int = 0


class FindingHierarchyResponse(BaseModel):
    assets: list[FindingHierarchyItem] = Field(default_factory=list)
    ungrouped: list[FindingHierarchyItem] = Field(default_factory=list)
    counts: FindingHierarchyCounts
    generated_at: datetime
    projection_version: str = "finding-hierarchy-v1"


class FindingDuplicateCandidate(BaseModel):
    id: UUID
    title: str
    target: str | None = None
    severity: Severity
    status: FindingStatus
    exclusion: FindingExclusion | None = None
    match_reason: str


class FindingFromHierarchyItemCreate(BaseModel):
    item_id: str = Field(min_length=8, max_length=80)
    title: str = Field(min_length=1, max_length=300)
    summary: str | None = Field(default=None, max_length=MAX_FINDING_SUMMARY_CHARS)
    severity: Severity = Severity.info
    phase: FindingPhase = FindingPhase.general
    target: str | None = Field(default=None, max_length=500)
    observed_at: datetime | None = None
    duplicate_decision: Literal["review", "create_anyway"] = "review"
    reviewed_duplicate_ids: list[UUID] = Field(default_factory=list, max_length=10)
    idempotency_key: UUID

    @field_validator("title")
    @classmethod
    def _title_not_blank(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("title cannot be blank")
        return normalized

    @field_validator("summary", "target")
    @classmethod
    def _normalize_optional(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class FindingFromHierarchyItemResponse(BaseModel):
    state: Literal["duplicate_warning", "created"]
    candidates: list[FindingDuplicateCandidate] = Field(default_factory=list)
    finding: FindingRead | None = None


FindingHierarchyItem.model_rebuild()

File: frontend/lib/finding-hierarchy.ts
import type {
  FindingHierarchyFindingRef,
  FindingHierarchyItem,
  FindingWorkspaceBucket,
  FindingWorkspaceView,
  Severity,
} from "@/lib/types";

export const FINDING_WORKSPACE_VIEWS: Array<{
  value: FindingWorkspaceView;
  label: string;
}> = [
  { value: "focus", label: "Focus" },
  { value: "needs_review", label: "Needs review" },
  { value: "actionable", label: "Actionable Findings" },
  { value: "inventory", label: "Inventory" },
  { value: "resolved_excluded", label: "Resolved / Excluded" },
];

export const FINDING_WORKSPACE_VIEW_VALUES = new Set<FindingWorkspaceView>(
  FINDING_WORKSPACE_VIEWS.map((view) => view.value),
);

const SEVERITY_RANK: Record<Severity, number> = {
  info: 0,
  low: 1,
  medium: 2,
  high: 3,
  critical: 4,
};

export function refMatchesView(
  ref: FindingHierarchyFindingRef,
  view: FindingWorkspaceView,
): boolean {
  if (view === "focus") {
    return ref.bucket === "needs_review" || ref.bucket === "actionable";
  }
  return ref.bucket === (view as FindingWorkspaceBucket);
}

function visibleRollup(item: FindingHierarchyItem): FindingHierarchyItem["rollup"] {
  const refs = new Map<string, FindingHierarchyFindingRef>();
  const collect = (node: FindingHierarchyItem) => {
    node.finding_refs.forEach((ref) => refs.set(ref.id, ref));
    node.children.forEach(collect);
  };
  collect(item);
  const values = Array.from(refs.values());
  const counts = {
    needs_review: 0,
    actionable: 0,
    inventory: 0,
    resolved_excluded: 0,
  };
  let maxSeverity: Severity = "info";
  let latestAt: string | null = null;
  values.forEach((ref) => {
    counts[ref.bucket] += 1;
    const observed = ref.observed_at ?? ref.created_at;
    if (latestAt === null || observed > latestAt) latestAt = observed;
    if (
      ref.bucket !== "resolved_excluded" &&
      SEVERITY_RANK[ref.severity] > SEVERITY_RANK[maxSeverity]
    ) {
      maxSeverity = ref.severity;
    }
  });
  return {
    max_severity: maxSeverity,
    distinct_findings: values.length,
    latest_at: latestAt,
    ...counts,
  };
}

function textMatches(item: FindingHierarchyItem, query: string): boolean {
  if (!query) return true;
  const fields = [
    item.label,
    item.value,
    item.ip,
    item.hostname,
    item.protocol,
    item.port?.toString(),
    item.service,
    item.url,
    ...item.finding_refs.flatMap((ref) => [
      ref.id,
      ref.title,
      ref.tool,
      ref.target,
    ]),
  ];
  return fields.some((field) => field?.toLowerCase().includes(query));
}

export function filterHierarchyItem(
  item: FindingHierarchyItem,
  view: FindingWorkspaceView,
  rawQuery = "",
): FindingHierarchyItem | null {
  const query = rawQuery.trim().toLowerCase();
  const directRefs = item.finding_refs.filter((ref) => refMatchesView(ref, view));
  const childMatches = item.children
    .map((child) => filterHierarchyItem(child, view, rawQuery))
    .filter((child): child is FindingHierarchyItem => child !== null);
  const directTextMatch = textMatches(item, query);
  const hasViewContent = directRefs.length > 0 || childMatches.length > 0;
  if (!hasViewContent) return null;
  if (query && !directTextMatch && childMatches.length === 0) return null;
  const filtered: FindingHierarchyItem = {
    ...item,
    finding_refs: directRefs,
    children:
      query && directTextMatch
        ? item.children
            .map((child) => filterHierarchyItem(child, view, ""))
            .filter((child): child is FindingHierarchyItem => child !== null)
        : childMatches,
  };
  filtered.rollup = visibleRollup(filtered);
  return filtered;
}

export function filterHierarchy(
  items: FindingHierarchyItem[],
  view: FindingWorkspaceView,
  search = "",
): FindingHierarchyItem[] {
  return items
    .map((item) => filterHierarchyItem(item, view, search))
    .filter((item): item is FindingHierarchyItem => item !== null)
    .sort((a, b) => {
      const review = Number(b.rollup.needs_review > 0) - Number(a.rollup.needs_review > 0);
      if (review !== 0) return review;
      const severity =
        SEVERITY_RANK[b.rollup.max_severity] -
        SEVERITY_RANK[a.rollup.max_severity];
      if (severity !== 0) return severity;
      if (b.rollup.actionable !== a.rollup.actionable) {
        return b.rollup.actionable - a.rollup.actionable;
      }
      return a.label.localeCompare(b.label);
    });
}
